app/merginapi.cpp

```diff
@@ -42,18 +114,122 @@ void MerginApi::listProjectsReplyFinished()
     if (r->error() == QNetworkReply::NoError)
     {
       QByteArray data = r->readAll();
+      cacheProjectsData(data);
       mMerginProjects = parseProjectsData(data);
     }
     else {
         QString message = QStringLiteral("Network API error: %1(): %2").arg("listProjects", r->errorString());
         qDebug("%s", message.toStdString().c_str());
-        emit networkErrorOccurred( r->errorString() );
+        emit networkErrorOccurred( r->errorString(), "Mergin API error: listProjects" );
     }
 
     r->deleteLater();
     emit listProjectsFinished();
 }
 
+void MerginApi::downloadProjectReplyFinished()
+{
+
+    QNetworkReply* r = qobject_cast<QNetworkReply*>(sender());
+    if (!r) {
+        emit networkErrorOccurred( "No reply received.", "Mergin API error: downloadProject" );
+        qDebug() << r->errorString();
+        mPendingRequests.remove(r->url());
+        return;
+    }
```

Comment:
I don't think it can ever happen that there's no reply. And even if it could, it would crash inside the block because of calls to r->url() but r is null ptr. I would suggest to replace this by Q_ASSERT(r);